share/blob.go

```diff
@@ -25,14 +25,25 @@ func NewBlob(ns Namespace, data []byte, shareVersion uint8, signer []byte) (*Blo
 	if len(data) == 0 {
 		return nil, errors.New("data can not be empty")
 	}
-	if shareVersion == 0 && signer != nil {
-		return nil, errors.New("share version 0 does not support signer")
+	if ns.IsEmpty() {
+		return nil, errors.New("namespace can not be empty")
 	}
-	if shareVersion == 1 && len(signer) != SignerSize {
-		return nil, fmt.Errorf("share version 1 requires signer of size %d bytes", SignerSize)
+	if ns.Version() != NamespaceVersionZero {
+		return nil, fmt.Errorf("namespace version must be %d got %d", NamespaceVersionZero, ns.Version())
 	}
-	if shareVersion > MaxShareVersion {
-		return nil, fmt.Errorf("share version can not be greater than MaxShareVersion %d", MaxShareVersion)
+	switch shareVersion {
+	case ShareVersionZero:
+		if signer != nil {
+			return nil, errors.New("share version 0 does not support signer")
+		}
+	case ShareVersionOne:
+		if len(signer) != SignerSize {
+			return nil, fmt.Errorf("share version 1 requires signer of size %d bytes", SignerSize)
+		}
+	// Note that we don't specifically check that shareVersion is less than 128 as this is caught
+	// by the default case
+	default:
+		return nil, fmt.Errorf("share version %d not supported. Please use 0 or 1", shareVersion)
```

Comment:
If the goal is to make the general constructor more restrictive by adding share version check, I propose to make it private in favour of keeping  `NewBlobV0/1` public only. It a nit but:
* It's less code and less test cases to cover.
* Its scales better as adding new versions requires less cognitive load to think where to add new cases for new versions. You simply add a new versioned constructor and that's it.

Reply:
You could also move namespace version checks to versioned constructors as that would simplify expressing relationships between namespace and share versions. E.g. BlobV3 does not work with NamespaceV0, but BlobV2 does.

Reply:
hmm I want to avoid this matrix of share and namespace versions

Reply:
The version matrix is irrelevant mostly. The point is about removing unnecessary checks for something that we already enforce with versioned constructors. I dont see much value in doing the same thing twice.

The matrix point is an additional thought, that in case it happens(i presumed so) its easier to handle in versioned constructor rather then general.

Reply:
We use `NewBlob` often when parsing from shares or from proto. If I were to make this private then I would then need to move the switch statements to these locations. I'm going to keep the `NewBlob` constructor as is